### PR TITLE
Fix mishandling of unions while freezing types in the mypy plugin

### PR DESCRIPTION
**Pull request by dmontagu**
This addresses the specific issue brought up in https://github.com/pydantic/pydantic/issues/7399, but I am concerned that there is a bigger problem that I'm not adequately addressing here.

I will raise an issue in the mypy repository asking for more information, as I think the problem that the pydantic mypy plugin is hitting has something to do with a warning comment in code I adapted from the dataclasses plugin.

If we are going to merge this, we should update the mypy plugin tests to include the example from #7399 (including adding the `unreachable` warning).

Closes https://github.com/pydantic/pydantic/issues/7399

### CODE REVIEW

**Comment by dmontagu** [comment]
I think if this still passes CI let's just merge it @sydney-runkle, I created the mypy issue https://github.com/python/mypy/issues/16087 but haven't heard anything back so I think it makes sense to address the case people are actually hitting and worry about doing a better job in the future if more issues get reported.

**Review by sydney-runkle — APPROVED** [review]
Nice. Waiting on approval of the `v2.5.0` release, then will merge this later today :).

**Pull request merged**
